micro-manager.py

```diff
@@ -119,42 +120,43 @@ def write_block_matrix_data(m, solver_interface, data_ids, vertex_ids):
 while interface.is_coupling_ongoing():
     # Write checkpoint
     if interface.is_action_required(precice.action_write_iteration_checkpoint()):
-        for v in range(nv):
-            micro_sims[v].save_state()
+        for v in range(nms):
+            micro_sims[v].save_checkpoint()
 
         t_checkpoint = t
         n_checkpoint = n
         interface.mark_action_fulfilled(precice.action_write_iteration_checkpoint())
 
-    # Read temperature values from preCICE
-    for data_id in readDataIDs:
-        readData = interface.read_block_scalar_data(data_id, macroVertexIDs)
-
-    print("Rank {} is solving micro simulations...".format(rank))
-    k, phi = [], []
-    i = 0
-    for data in readData:
-        micro_sims[i].refine_mesh()
-        phi_i = micro_sims[i].solve_allen_cahn(temperature=data, dt=dt)
-        phi.append(phi_i)
+    for name, dims in read_data_names.items():
+        if dims == 0:
+            read_data.update({name: interface.read_block_scalar_data(read_data_ids[name], mesh_vertex_ids)})
+        elif dims == 1:
+            read_data.update({name: interface.read_block_vector_data(read_data_ids[name], mesh_vertex_ids)})
 
-        k_i = micro_sims[i].solve_heat_cell_problem()
-        k.append(k_i)
+    micro_sims_input = [dict(zip(read_data, t)) for t in zip(*read_data.values())]
+    micro_sims_output = []
+    for i in range(nms):
+        micro_sims_output.append(micro_sims[i].solve(micro_sims_input[i], dt))
 
-        i += 1
+    write_data = {k: [d[k] for d in micro_sims_output] for k in micro_sims_output[0]}
```

Comment:
@Scriptkiddi I really like these one liners that you suggested, but this specific one is causing some problems. The end result here is a dictionary `write_data` but the values of the dictionary end up being several lists embedded in a list. So for example:

```
write_data["porosity"] = [[0.5], [0.73], [0.78], ...]
```
where the individual lists are values acquired from each micro-simulation. What we need here is one list which is appended with all the values. I tried to get it to work that way but could not get it done. Any ideas?

Reply:
We should enforce that scalars and vectors are always wrapped in a list when returned by the micro simulation
than we can use `{k: functools.reduce(operator.iconcat, [dic[k] for dic in a], []) for k in a[0]}`  to flatten the lists for the write_data list